yaksh/models.py

```diff
@@ -1108,6 +1133,112 @@ def has_lessons(self):
+        model_class = get_model_class(model_type)
+        new_obj, obj_create_status = model_class.objects.get_or_create(**model_obj)
+        return new_obj, obj_create_status
+
+    @classmethod
+    def load_yaml(cls, filepath, user):
+        file_list = {}
+        with zipfile.ZipFile(filepath,'r') as zfile:
+            for name in zfile.namelist():
+                file_content = zfile.read(name)
+                yaml_content = ruamel.yaml.safe_load(file_content)
+                if yaml_content.get('type') and yaml_content.get('type') == 'course':
+                    course = yaml_content
+                file_list.update(
+                    {name: yaml_content}
+                )
+        subitems = course.pop('subitems')
+        model_type = course.pop('type')
+        new_course, course_create_status = cls._create_object_from_dict(course, 'course')
+        new_mod_list = []
+        for mod in subitems:
+            mod_dict = file_list[mod]
+            mod_sub_items = mod_dict.pop('subitems')
+            model_type = mod_dict.pop('type')
+            mod_dict.update({'creator': user})
+            new_mod, mod_create_status = cls._create_object_from_dict(mod_dict, 'learningmodule')
+            new_mod_list.append(new_mod)
+            new_unit_list = []
+            for unit in mod_sub_items:
+                unit_dict = file_list[unit]
+                unit_sub_items = unit_dict.pop('subitems')
+                model_type = unit_dict.pop('type')
+                lesson_data = unit_sub_items[0]
+                lesson_dict = file_list[lesson_data]
+                lesson_dict.pop('subitems')
+                lesson_dict.pop('type')
+                lesson_dict.update({'creator': user})
+                new_lesson, lesson_create_status = cls._create_object_from_dict(lesson_dict, 'lesson')
+                new_unit, unit_create_status = LearningUnit.objects.get_or_create(lesson=new_lesson, type='lesson', **unit_dict)
+                new_unit_list.append(new_unit)
+            new_mod.learning_unit.add(*new_unit_list)
+        new_course.learning_module.add(*new_mod_list)
+
+    def create_yaml(self, path=None):
+        if path:
+            dir_path = path
+        else:
+            dir_path = tempfile.mkdtemp()
+
+        zip_path = os.path.join(
+            dir_path, '{0}.zip'.format(slugify(self.name))
+        )
+        with zipfile.ZipFile(zip_path, "a") as zip_file:
+            fields = [
+                'name', 'enrollment', 'active', 'code',
+                'instructions', 'view_grade'
+            ]
+
+            mod_subitems = []
+            for module in self.get_learning_modules():
+                mod_fields = [
+                    'name', 'description', 'check_prerequisite',
+                    'check_prerequisite', 'check_prerequisite_passes',
+                    'html_data', 'active'
+                ] 
+
+                unit_subitems = []
+                for unit in self.get_learning_units():
```

Comment:
Shouldn't this be `module.get_learning_units()`?

Reply:
Using the course instance method with the same name.